### PR TITLE
compare scaffold folders instead of rebuilding

### PR DESCRIPTION
**Pull request by Geal**
This adds to the repository the files that the scaffold test is supposed to generate, makes sure they are built and tested as part of the workspace, then when the scaffold test runs, instead of rebuilding a router from scratch, we only compare that the generated folder is the same. To that end, we skip the generated `Cargo.toml` files because they will contain hardcoded paths dependent on the repository's path

*Description here*

Fixes #**issue_number**

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

**Comment by apollo-bot2** [comment]
## Detected SAST Vulnerabilities
## :red_circle: Potential Security Issues Found

If you are seeing this message, it means that [the security scanning tool](https://semgrep.dev/) that Apollo uses to test our source has identified a potential security issue in code introduced or updated via your branch. Information about what was detected and steps to move forward are below.

If the scanner detected a legitimate issue, please take action to correct it prior to merging this PR. The action required will vary based on the detection. If the detection is a false positive, please follow the steps below to resolve the issue.

###  Issue Detected 
| Message | Caused CI Failure? | Link | Path | Signature |
|---------|--------------------|------|------|-----------|
| By not specifying a USER, a program in the container may run as 'root'. This is a security hazard. If an attacker can control a process running as root, they may have control over the container. Ensure that the last USER in a Dockerfile is a USER other than 'root'. | Yes | <a href="https://github.com/apollographql/router/blob/f1ec438b8d5fac1dfc46a89a44dcb54f875eb849/apollo-router-scaffold/scaffold-test/Dockerfile?plain=1#L49" target="_blank">Link</a> | apollo-router-scaffold/scaffold-  test/Dockerfile | rules.providers.semgrep.security.do  ckerfile.security.missing-user |
<hr />
<details>
    <summary><h3>False Positive Resolution Process</h3></summary>

The easiest way to resolve a false positive is to add a comment containing `nosemgrep` to the code that triggered the detection. You can add the comment directly to the line triggering the detection or on the line immediately above it. For example:

```python
#!/usr/bin/env python3

def my_function():
    my_code_generating_detection() # nosemgrep
```

<details>
    <summary><h4>What if I can't add a line comment?</h4></summary>

If you can't add a comment because the detection is generated by a file that does not support comments (like JSON), you can use Semgrep's `.semgrepignore` file. For more information about how to create an exclusion via `.semgrepignore`, refer to the [Semgrep documentation](https://semgrep.dev/docs/ignoring-files-folders-code/#defining-ignored-files-and-folders-in-semgrepignore).
</details>

</details>
<hr />
<details> 
    <summary><h3>Getting Help</h3></summary>


The Apollo Security team is available to assist in resolving this issue. Please tag us on this PR using  `@apollographql/security` if you need assistance! 
</details>
<hr />
<details>
    <summary><h3>How do I know I fixed this correctly?</h3></summary>
Once you have resolved the issue completely, this message will disappear! If you're still seeing this message, there is more to do prior to merging.
</details>

**Review by bnjjj — APPROVED** [review]

File: apollo-router-scaffold/templates/plugin/src/plugins/{{snake_name}}.rs [hunk 1]
```diff
@@ -1,26 +1,33 @@
...
-use std::ops::ControlFlow;
-use tower::ServiceExt;
 use tower::ServiceBuilder;
+use tower::ServiceExt;
 {{/if}}
 {{#if type_tracing}}
-use apollo_router::layers::ServiceBuilderExt;
-use tower::ServiceExt;
 use tower::ServiceBuilder;
+use tower::ServiceExt;
```
Comment (SimonSapin) [inline apollo-router-scaffold/templates/plugin/src/plugins/{{snake_name}}.rs:29]: Are these changes intentional? They look possibly made by a formatter that does understand template conditionnals

**Review by SimonSapin — APPROVED** [review]

File: apollo-router-scaffold/scaffold-test/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+# renovate-automation: rustc version
+FROM rust:1.76.0 as build
+
+# Set our working directory for the build
+WORKDIR /usr/src/router
+
+# Update our build image and install required packages
+RUN apt-get update
+RUN apt-get -y install \
+    npm \
```
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router-scaffold/scaffold-test/Dockerfile:11]: probably. I have not rewritten the templated Dockerfile to follow our latest version

File: apollo-router-scaffold/scaffold-test/README.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+1. Download the example schema
+
+   ```bash
+   curl -sSL https://supergraph.demo.starstuff.dev/ > supergraph-schema.graphql
+   ```
+
+2. Run the Apollo Router
+
+   During development it is convenient to use `cargo run` to run the Apollo Router as it will
```
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router-scaffold/scaffold-test/README.md:32]: looks like it, but that was in the original template https://github.com/apollographql/router/blob/dev/apollo-router-scaffold/templates/base/README.md

File: apollo-router-scaffold/templates/plugin/src/plugins/{{snake_name}}.rs (hunk 1 above)
Comment (Geal) [reply apollo-router-scaffold/templates/plugin/src/plugins/{{snake_name}}.rs:29]: yes, the formatter that understands template conditional was me :sweat_smile: 
The generated project that is now committed would be flagged by the lint step, the `ignore` option in `rustfmt.toml` is not stable yet, so I chose to format the project, then edit the template so it generates properly formatted code
